language/move-prover/boogie-backend/src/boogie_helpers.rs

```diff
@@ -189,78 +191,180 @@ pub fn boogie_type_value_array(env: &GlobalEnv, args: &[Type]) -> String {
 
 /// Creates a type value array for types given as strings.
 pub fn boogie_type_value_array_from_strings(args: &[String]) -> String {
+    boogie_make_vec_from_strings(args)
+}
+
+/// Creates a vector from the given list of arguments.
+pub fn boogie_make_vec_from_strings(args: &[String]) -> String {
     if args.is_empty() {
-        return "$EmptyTypeValueArray".to_string();
-    }
-    let mut map = String::from("$MapConstTypeValue($DefaultTypeValue())");
-    for (i, arg) in args.iter().enumerate() {
-        map = format!("{}[{} := {}]", map, i, arg);
+        "EmptyVec()".to_string()
+    } else {
+        let n = usize::min(args.len(), MAX_MAKE_VEC_ARGS);
+        let direct_args = &args[0..n];
+        let mut make = format!("MakeVec{}({})", n, direct_args.iter().join(", "));
+        for arg in args.iter().skip(n) {
+            make = format!("ExtendVec({}, {})", make, arg)
+        }
+        make
     }
-    format!("$TypeValueArray({}, {})", map, args.len())
 }
 
 /// Return boogie type for a local with given signature token.
 pub fn boogie_local_type(ty: &Type) -> String {
-    if ty.is_reference() {
-        "$Mutation".to_string()
-    } else {
-        "$Value".to_string()
+    use PrimitiveType::*;
+    use Type::*;
+    match ty {
+        Primitive(p) => match p {
+            U8 | U64 | U128 | Num | Address | Signer => "int".to_string(),
+            Bool => "bool".to_string(),
+            TypeValue => "$TypeValue".to_string(),
+            _ => panic!("unexpected type"),
+        },
+        Vector(..) | Struct(..) => "Vec $Value".to_string(),
+        Reference(..) => "$Mutation".to_string(),
+        TypeParameter(..) => "$Value".to_string(),
+        Fun(..) | Tuple(..) | TypeDomain(..) | ResourceDomain(..) | TypeLocal(..) | Error
+        | Var(..) => format!("<<unsupported: {:?}>>", ty),
+    }
+}
+
+pub fn boogie_declare_temps() -> String {
+    "var $$t_bool: bool;\nvar $$t_int: int;\nvar $$t_addr: int;\nvar $$t_type: $TypeValue;\
+    \nvar $$t, $$t1, $$t2, $$t3: $Value;\nvar $$t_vec: Vec $Value;"
+        .to_string()
+}
+
+pub fn boogie_temp(ty: &Type) -> String {
+    format!("$$t{}", boogie_type_suffix(ty))
+}
+
+pub fn boogie_type_suffix(ty: &Type) -> &'static str {
+    use PrimitiveType::*;
+    use Type::*;
+    match ty {
+        Primitive(p) => match p {
+            U8 | U64 | U128 | Num | Signer => "_int",
+            Address => "_addr",
+            Bool => "_bool",
+            TypeValue => "_type",
+            EventStore => "",
+            _ => "<<unsupported>>",
+        },
+        Vector(..) | Struct(..) => "_vec",
+        TypeParameter(..) => "",
+        Fun(..) | Tuple(..) | TypeDomain(..) | ResourceDomain(..) | TypeLocal(..) | Error
+        | Var(..) | Reference(..) => "<<unsupported>>",
```

Comment:
Just curious whether it will be good to `panic` here instead of outputting an `<<unsupported>>` tag? (and in couple of other places too)

Reply:
Actually the "unsupported" makes it a bit easier to debug, because the subsequent failure happens in Boogie, which is easier to debug, as you can see where it happens. Even better would be to report a regular `env.error` on Move code, but right now there is no env here.